lib/internal/promisify.js

```diff
@@ -0,0 +1,43 @@
+'use strict';
+
+module.exports = promisify;
+
+function promisify(fn, single) {
+  Object.defineProperty(inner, 'name', {
+    value: fn.name + 'Promisified'
+  });
+  return inner;
+
+  function inner() {
+    if (typeof arguments[arguments.length - 1] === 'function') {
+      return fn.apply(this, arguments);
+    }
+    var resolve = null;
+    var reject = null;
+    const args = Array.from(arguments);
```

Comment:
> Function can be already deopted because of try-catch below, so ¯\_(ツ)_/¯

Seems like it's not an issue anymore with the latest V8: https://twitter.com/kdzwinel/status/667078434598682624

>  function is already just a wrapper for another one

And this is exactly what bothers me: it's used as a wrapper for all promised API. But, as someone pointed above, IO is so extremely slow that this deopt may not affect overall performance. So more likely you're right, but it would be nice to do at least some basic benchmarking.